server/automute_test.go

```diff
@@ -0,0 +1,451 @@
+	for _, preference := range preferences {
+		a.preferences[a.key(userID, preference.Category, preference.Name)] = preference
+	}
+
+	return nil
+}
+
+func (a *AutomuteAPIMock) CreateChannel(channel *model.Channel) (*model.Channel, *model.AppError) {
+	a.channels[channel.Id] = channel
+	return channel, nil
+}
+
+func (a *AutomuteAPIMock) GetDirectChannel(userID1, userID2 string) (*model.Channel, *model.AppError) {
+	channel, _ := a.CreateChannel(&model.Channel{
+		Id:   model.NewId(),
+		Type: model.ChannelTypeDirect,
+	})
+	a.AddUserToChannel(channel.Id, userID1, "")
+	a.AddUserToChannel(channel.Id, userID2, "")
+	return channel, nil
+}
+
+func (a *AutomuteAPIMock) AddUserToChannel(channelID, userID, asUserID string) (*model.ChannelMember, *model.AppError) {
+	a.channelMembers[a.key(channelID, userID)] = &model.ChannelMember{
+		UserId:      userID,
+		ChannelId:   channelID,
+		NotifyProps: model.GetDefaultChannelNotifyProps(),
+	}
+	return a.channelMembers[a.key(channelID, userID)], nil
+}
+
+func (a *AutomuteAPIMock) GetChannelsForTeamForUser(teamID, userID string, includeDeleted bool) ([]*model.Channel, *model.AppError) {
+	if teamID != "" || !includeDeleted {
+		panic("Not implemented")
+	}
+
+	var channels []*model.Channel
+	for _, channelMember := range a.channelMembers {
+		if channelMember.UserId != userID {
+			continue
+		}
+
+		channels = append(channels, a.channels[channelMember.ChannelId])
+	}
+	return channels, nil
+}
+
+func (a *AutomuteAPIMock) GetChannelMember(channelID, userID string) (*model.ChannelMember, *model.AppError) {
+	member, ok := a.channelMembers[a.key(channelID, userID)]
+	if !ok {
+		return nil, &model.AppError{Message: "Channel member not found"}
+	}
+	return member, nil
+}
+
+func (a *AutomuteAPIMock) PatchChannelMembersNotifications(identifiers []*model.ChannelMemberIdentifier, notifyProps map[string]string) *model.AppError {
+	for _, identifier := range identifiers {
+		for propKey, propValue := range notifyProps {
+			a.channelMembers[a.key(identifier.ChannelId, identifier.UserId)].NotifyProps[propKey] = propValue
+		}
+	}
+
+	return nil
+}
+
+func (a *AutomuteAPIMock) LogDebug(msg string, keyValuePairs ...any) {
+	a.t.Log(msg, keyValuePairs)
+}
+
+func (a *AutomuteAPIMock) LogInfo(msg string, keyValuePairs ...any) {
+	a.t.Log(msg, keyValuePairs)
+}
+
+func (a *AutomuteAPIMock) LogError(msg string, keyValuePairs ...any) {
+	a.t.Error(msg, keyValuePairs)
+}
+
+func (a *AutomuteAPIMock) LogWarn(msg string, keyValuePairs ...any) {
+	a.t.Log(msg, keyValuePairs)
+}
+
+func newAutomuteTestPlugin(t *testing.T) *Plugin {
+	mockAPI := &AutomuteAPIMock{
+		API:            &plugintest.API{},
+		channels:       make(map[string]*model.Channel),
+		channelMembers: make(map[string]*model.ChannelMember),
+		preferences:    make(map[string]model.Preference),
+		t:              t,
+	}
+	t.Cleanup(func() { mockAPI.AssertExpectations(t) })
+
+	p := newTestPlugin(t)
+	p.SetAPI(mockAPI)
+
+	return p
+}
```

Comment:
I would love to not need this, but I guess is the state we are at with plugin testing.

Reply:
Yeah, same. Unless we run the tests with an actual server or restructure our code in some way that we use more pure functions and less side effects, we're stuck mocking something.

Thinking about this more since last week though, I do prefer this style of mocking where you make a fake server once and then can use it all over instead of faking return values for individual API methods like we've had to do elsewhere